edgehog-device-runtime-forwarder/src/messages.rs

```diff
@@ -0,0 +1,468 @@
+// This file is part of Edgehog.
+//
+// Copyright 2023 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Internal Rust representation of protobuf structures.
+//!
+//! The structures belonging to this module are used to serialize/deserialize to/from the protobuf
+//! data representation.
+
+use std::borrow::{Borrow, BorrowMut};
+use std::collections::HashMap;
+use std::fmt::{Debug, Display, Formatter};
+use std::num::TryFromIntError;
+use std::ops::{Deref, DerefMut, Not};
+use std::str::FromStr;
+
+use displaydoc::Display;
+use http::{HeaderMap, HeaderName, HeaderValue};
+use prost::Message as ProstMessage;
+use reqwest::{Client as ReqwClient, RequestBuilder};
+use thiserror::Error as ThisError;
+use url::ParseError;
+
+use edgehog_device_forwarder_proto as proto;
+use edgehog_device_forwarder_proto::{
+    http::Message as ProtoHttpMessage, http::Request as ProtoHttpRequest,
+    http::Response as ProtoHttpResponse, message::Protocol as ProtoProtocol,
+    web_socket::Message as ProtoWsMessage, WebSocket as ProtoWebSocket,
+};
+
+/// Errors occurring while handling [`protobuf`](https://protobuf.dev/overview/) messages
+#[derive(Display, ThisError, Debug)]
+#[non_exhaustive]
+pub enum ProtoError {
+    /// Failed to serialize into Protobuf.
+    Encode(#[from] prost::EncodeError),
+    /// Failed to deserialize from Protobuf.
+    Decode(#[from] prost::DecodeError),
+    /// Empty fields.
+    Empty,
+    /// Reqwest error.
+    Reqwest(#[from] reqwest::Error),
+    /// Error parsing URL.
+    ParseUrl(#[from] ParseError),
+    /// Wrong HTTP method field.
+    InvalidHttpMethod(#[from] http::method::InvalidMethod),
+    /// Http error.
+    Http(#[from] http::Error),
+    /// Error while parsing Headers.
+    ParseHeaders(#[from] http::header::ToStrError),
+    /// Invalid port number
+    InvalidPortNumber(#[from] TryFromIntError),
+}
+
+/// Requests Id.
+#[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct Id(Vec<u8>);
```

Comment:
Is it really necessary to define this struct and its methods and not use only `Vec<u8>`?

Reply:
I think it is better to implement it. It's necessary to implement Debug with `hex` output and it's a clearer name (for instance, instead of passing `Vec<u8>` to all functions). Also it is easier to refactor in the future.